core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -628,6 +621,13 @@ private[spark] class ExecutorAllocationManager(
             allocationManager.onExecutorIdle(executorId)
           }
         }
+
+        // If the task failed, we expect it to be resubmitted later.
+        if (taskEnd.reason != Success) {
+          stageIdToTaskIndices.get(stageId).foreach { taskIndices =>
+            taskIndices.remove(taskIndex)
```

Comment:
OK -- I should add that I'm a little out of my depth here, I will defer the final review to andrew, just trying to speed things up if there is anything missed.  Sorry if I'm just adding noise